File: day01/src/main/java/de/havox_design/aoc2016/day01/Direction.java
package de.havox_design.aoc2016.day01;

public enum Direction {
    NORTH,
    SOUTH,
    EAST,
    WEST;

    public Direction left() {
        return switch (this) {
            case NORTH -> WEST;
            case SOUTH -> EAST;
            case EAST -> NORTH;
            case WEST -> SOUTH;
        };
    }

    public Direction right() {
        return switch (this) {
            case NORTH -> EAST;
            case SOUTH -> WEST;
            case EAST -> SOUTH;
            case WEST -> NORTH;
        };
    }

    public Point next(Point point, int offset) {
        return switch (this) {
            case NORTH -> new Point(point.x(), point.y() + offset);
            case SOUTH -> new Point(point.x(), point.y() - offset);
            case EAST -> new Point(point.x() + offset, point.y());
            case WEST -> new Point(point.x() - offset, point.y());
        };
    }
}

File: aoc2024/aoc2024-kotlin/src/main/kotlin/de/havox_design/aoc2024/day04/Direction.kt
package de.havox_design.aoc2024.day04

enum class Direction(val dx: Int, val dy: Int) {
    EAST(1, 0),
    NORTHEAST(1, -1),
    NORTH(0, -1),
    NORTHWEST(-1, -1),
    WEST(-1, 0),
    SOUTHWEST(-1, 1),
    SOUTH(0, 1),
    SOUTHEAST(1, 1),
    ;

    operator fun component1() = dx
    operator fun component2() = dy
}
